File: go/bob/bob.go
package bob // package name must match the package name in bob_test.go
import "unicode"

const testVersion = 2 // same as targetTestVersion

const shoutResponse = "Whoa, chill out!"
const questionResponse = "Sure."
const nonsenseResponse = "Fine. Be that way!"
const defaultResponse = "Whatever."

func Hey(s string) string {
	lastChar := string(s[len(s)-1])
	allCaps := true

	for _, char := range s {
		if unicode.IsLetter(char) {
			if !unicode.IsUpper(char) {
				allCaps = false
			}
		}
	}
	if allCaps {
		return shoutResponse
	} else if lastChar == "?" {
		return questionResponse
	}

	return defaultResponse
}

File: elixir/bob/bob.exs
defmodule Bob do
  @upper_case_regex ~r/^[A-Z]*$/
  @alpha_regex ~r/[^A-Za-z]/
  @silence_regex ~r/^[\W]*$/

  def hey(input) do
    cond do
      silence?(input) -> "Fine. Be that way!"
      all_upper_case?(input) && is_question?(input) -> "Calm down, I know what I'm doing!"
      all_upper_case?(input) -> "Whoa, chill out!"
      is_question?(input) -> "Sure."
      true -> "Whatever."
    end
  end

  defp all_upper_case?(input) do
    input
    |> String.replace(@alpha_regex, "")
    |> String.match?(@upper_case_regex)
  end

  defp is_question?(input), do: String.ends_with?(input, "?")

  defp silence?(input), do: String.match?(input, @silence_regex)
end
